controller/internal/enforcer/dnsproxy/dns.go

```diff
@@ -0,0 +1,225 @@
+package dnsproxy
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.aporeto.io/trireme-lib/collector"
+	"go.aporeto.io/trireme-lib/controller/pkg/flowtracking"
+	"go.aporeto.io/trireme-lib/controller/pkg/pucontext"
+	"go.aporeto.io/trireme-lib/policy"
+	"go.aporeto.io/trireme-lib/utils/cache"
+	"go.uber.org/zap"
+)
+
+// Proxy struct represents the object for dns proxy
+type Proxy struct {
+	puFromID          cache.DataStore
+	conntrack         flowtracking.FlowClient
+	collector         collector.EventCollector
+	contextIDToServer map[string]*dns.Server
+	chreports         chan dnsReport
+	sync.RWMutex
+}
+
+type serveDNS struct {
+	contextID string
+	*Proxy
+}
+
+const (
+	dnsRequestTimeout = 2 * time.Second
+	proxyMarkInt      = 0x40 //Duplicated from supervisor/iptablesctrl refer to it
+)
+
+func socketOptions(_, _ string, c syscall.RawConn) error {
+	var opErr error
+	err := c.Control(func(fd uintptr) {
+		if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+			zap.L().Error("Failed to mark connection", zap.Error(err))
+		}
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return opErr
+}
+
+func listenUDP(network, addr string) (net.PacketConn, error) {
+	var lc net.ListenConfig
+
+	lc.Control = socketOptions
+
+	return lc.ListenPacket(context.Background(), network, addr)
+}
+
+func forwardDNSReq(r *dns.Msg, ip net.IP, port uint16) (*dns.Msg, []string, uint32, error) {
+	var ips []string
+	c := new(dns.Client)
+	c.Dialer = &net.Dialer{
+		Control: func(_, _ string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+					zap.L().Error("Failed to assing mark to socket", zap.Error(err))
+				}
+			})
+		},
+		Timeout: dnsRequestTimeout,
```

Comment:
Does the mark work here? Or do you need an external service anyway? I have a feeling that marking a UDP socket doesn't work.